core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java

```diff
@@ -63,8 +65,8 @@ public ProjectJoinTransposeRule(
       PushProjector.ExprCondition preserveExprCondition,
       RelBuilderFactory relFactory) {
     super(
-        operand(Project.class,
-            operand(Join.class, any())),
+        operand(LogicalProject.class,
+            operand(LogicalJoin.class, any())),
```

Comment:
With this matching pattern changed, do we still need below change -- the way how join node is created.

Reply:
Agree. May better keep the below part unchanged.

Reply:
I think the way to go is to allow clients to specify their own `Project`, `Join` and `RelBuilderFactory` classes. I guess our default `INSTANCE` of this rule should be adapted to match only `LogicalProject` and `LogicalJoin` and the Javadoc should be changed accordingly.